### PR TITLE
Modernise the table design

### PR DESCRIPTION
**Pull request by janfaracik**
<details>
  <summary>📸 View screenshots</summary>

##### Old
image

##### New
image

image

##### Old
image

##### New
image

image

##### Old

image

##### New
image

image
</details>

### What's changed and what's the reasoning behind it

**This MR updates the design of the table component in Jenkins, offering a more modern appearance that's easier to use and is easier to work with.**
> This is a large change and doesn't change every table in Jenkins, much like the forms changes I want to do this in sequences to minimise impact.
>
> As always not everything may work perfectly from the get go - I'd really appreciate any feedback/comments about it :) 

**This MR also adds an `ionicon` tag - [Ionicons](https://ionic.io/ionicons) is a wonderful icon library that I think would look great in Jenkins - it's open source, MIT licensed and super easy to integrate with.** 

> **Why not use the icons/SVGs that are there?**
> I found that they're rather difficult to work with, the old SVGs (previously PNGs) look old and antiquated and using the `icon` component with its class size system felt odd in an age where fully scalable graphics are a thing. 
>
> `svgIcon` was an improvement, but it required that all SVGs have an ID (an annoyance as that may require tweaking of the original SVG file) and the SVGs were all located in one file, which made finding appropriate icons tedious.
>
> My approach for Ionicons was to make using them super simple, simply write `<u:ionicon name="terminal-outline" title="${%Console output}" />` with the name and title you want for the icon and it'll appear on the page in SVG format. No need to worry about sizes (that'd be managed by the icons container) or IDs.
>
> NOTE: Jelly really isn't my forte, there might be an easier way of importing icons - if so please let me know and I'll make the changes :) 

The eventual plan if this MR is approved/merged would be to update the rest of the tables in Jenkins to follow this format, and then to update instances of icons to use Ionicons, eventually deprecating older styles/icon sources.

## Test this pull request

```command
docker run --rm -ti -p 8080:8080 -e ID=5851 -e MERGE_WITH=master jenkins/core-pr-tester
```

---

### Proposed changelog entries

* Modernise the table design and add support for Ionicons.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by uhafner — COMMENTED** [review]
I haven't looked at the code. But the new design is awesome! It is much cleaner and very modern. Additionally, using the new icons improves the old design even more: it is also very helpful that the SVG icons will derive the color from a variable so that they automatically show up right in different themes.

**Review by fqueiruga — APPROVED** [review]
This is an AMAZING PR @janfaracik . By creating a new table system and not limit yourself to `.bigtable` like I did in the past you really did achieve a lot of changes.

I have a few comments but most of them are related to SCSS codestyle. This is really great.

**Review by jtnord — COMMENTED** [review]
I am concerned that this requires changes in every plugin that uses a table, or at least that contributes columns to a table that is defined by core.
Have I missunderstood the changes here?

File: war/src/main/less/modules/tabs.less
Comment (janfaracik) [reply war/src/main/less/modules/tabs.less]: Removed them and moved the dark mode styling to the plugin :)

File: war/src/main/less/modules/tabs.less
Comment (janfaracik) [reply war/src/main/less/modules/tabs.less]: Removed them and moved the dark mode styling to the plugin :)

**Comment by janfaracik** [comment]
> Haven't looked much will do some more later,
> 
> Build history table: image
> 
> Pipeline syntax icon looks blurry or at least artifacts: image
> 
> Views icon is weird if I am using the default theme but have a dark system theme: image
> 
> image
> 
> Looks nice in actual dark theme: image
> 
> Although the dark theme needs some changes to adapt to new table: image

Updated the build history widget title, should be sorted now.

Where might I find the pipeline syntax link?

As for the colours, the light/dark mode bits were left in by mistake - just makes it easier to develop the light/dark mode version of the table/tabs :)

**Comment by timja** [comment]
Pipeline syntax is in the sidebar of a pipeline job

**Comment by timja** [comment]
@uhafner / @janfaracik 

Found this weirdness with columns and warnings-ng:

https://user-images.githubusercontent.com/21194782/139811574-fd388a89-2cb7-43c2-ae81-e6f7a3bdcfc1.mov

The column looks fine till you hover over it and then instead of a tooltip it's rendered inside the column. Tbh it loooks better than the offscreen tooltip you currently get, I'm just not sure if this is intentional.

The footer from nodes and clouds looks a bit funny on a wide screen:

last checked is mis-aligned, looking at the dom I think there's some more styles applied to the tbody area.

image

Generally looks really good and I think quite close, (need to sort the build though, I'll have a peek)

**Comment by uhafner** [comment]
I already fixed that. I just need to make a new release today...

**Comment by fqueiruga** [comment]
@timja about the issue with that footer, IIRC the markup for that table was a bit weird and I had a bad time styling it. It's a bit of an outlier and maybe can be revamped.

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:87]: So this fixes the tests (loading from classpath) but then breaks it from `jetty:run`:

```
            InputStream inputStream = getClass().getResourceAsStream("/images/ionicons/" + name + ".svg");
            if (inputStream == null) {
                return getIonicon("ellipse-outline", "Placeholder");
            }

            String ionicon = IOUtils.toString(inputStream, StandardCharsets.UTF_8);
```

This will only work from core as well, needs a look to see more how the rest of the Icon class fits together

File: core/src/main/resources/hudson/views/LastStableColumn/column.jelly
Comment (timja) [reply core/src/main/resources/hudson/views/LastStableColumn/column.jelly:32]: warnings ng looks fine

File: war/src/main/less/abstracts/theme.less
Comment (timja) [inline war/src/main/less/abstracts/theme.less:172]: have you got some working vars for dark theme you could submit a PR for?

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:98]: Not sure if there's a better way to load a resource when running with `jetty:run` but they don't appear to be available through the regular classpath location, these two paths are based on running from either the root directory of a jenkins checkout or the war sub folder

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:74]: From what I can tell this means that icons will only have a title for screen readers now, is that intentional / what we want?

i.e. if I hover over build now nothing shows up, whereas if I hover over it on another Jenkins instance there's a title that pops up

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:98]: I'll try moving them to src/main/resources

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:76]: Are plugins likely going to want to add their own ionicon's? if so there will need to be another method which takes a `Class` or `Classloader` as an argument so as to locate the correct resource I believe.

(This can probably be a follow up)

**Comment by timja** [comment]
> NOTE: Jelly really isn't my forte, there might be an easier way of importing icons - if so please let me know and I'll make the changes :)

I couldn't see a need for the custom tag lib so I changed it to use the one the other icons are using, feel free to revert or we can discuss :)

[`d809e85` (#5851)](https://github.com/jenkinsci/jenkins/pull/5851/commits/d809e8553526061673141f8795da264dc1413db8)

File: core/src/main/resources/lib/layout/ionicon.jelly
Comment (uhafner) [inline core/src/main/resources/lib/layout/ionicon.jelly:17]: This looks somehow strange that we need to invoke sever code to render an image. I'm using a SVG sprite to render Font Awesome icons. 

How can I add an additional `class` to the icon? E.g., to change color or size? 

I used that approach in the font-awesome plugin:
https://github.com/jenkinsci/font-awesome-api-plugin/blob/master/src/main/resources/font-awesome/svg-icon.jelly

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (uhafner) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:71]: I'm using the `<title>` tag for the Font Awesome icons: https://github.com/jenkinsci/font-awesome-api-plugin/blob/master/src/main/resources/font-awesome/svg-icon.jelly

File: core/src/main/resources/hudson/views/JobColumn/column.jelly
Comment (uhafner) [inline core/src/main/resources/hudson/views/JobColumn/column.jelly:29]: Can't we reuse the existing classes? Otherwise we need to change every plugin to get it adapted to the new style.

File: core/src/main/resources/hudson/views/LastStableColumn/column.jelly
Comment (uhafner) [reply core/src/main/resources/hudson/views/LastStableColumn/column.jelly:32]: It does not break, yes. But it uses the old styles. Wouldn't it be better to automatically get the new design for all?

File: core/src/main/resources/lib/hudson/buildHealth.jelly
Comment (uhafner) [inline core/src/main/resources/lib/hudson/buildHealth.jelly:70]: Also here: we need to restyle all existing plugins. Isn't there a better alternative without changing the plugins by substituting the existing classes?

File: war/src/main/less/abstracts/theme.less
Comment (uhafner) [inline war/src/main/less/abstracts/theme.less:314]: I will bring up that topic in a separate mail (see UX meeting) but wouldn't it make more sense to use some kind of semantics color names? Why is this green different from the success green?

File: war/src/main/less/modules/table.less
Comment (uhafner) [reply war/src/main/less/modules/table.less:16]: Shouldn't we then fix those tables?

File: war/src/main/less/modules/table.less
Comment (uhafner) [reply war/src/main/less/modules/table.less:100]: As a non UI developer I also find this new style syntax complex to understand. I have no idea how I can apply those new styles for the Bootstrap or DataTables plugins 🤷

File: war/src/main/resources/images/ionicons/ellipse-outline.svg
Comment (uhafner) [inline war/src/main/resources/images/ionicons/ellipse-outline.svg:1]: How can we use the other icons? So we need to manually add them one by one? Wouldn't it make more sense to include all (like in the Font Awesome plugin)?

File: core/src/main/resources/hudson/views/LastStableColumn/column.jelly
Comment (uhafner) [reply core/src/main/resources/hudson/views/LastStableColumn/column.jelly:32]: Here an example from the dashboard view:

Bildschirmfoto 2021-11-04 um 17 31 52

The upper part reuses the job table and looks good. The lower part uses the Bigtable CSS which seems to be unchanged. Also my newly introduced tooltips (for the tooltip PR) seem to be not automatically in synch with the new design .

**Review by uhafner — APPROVED** [review]
The new design is fantastic. It gives Jenkins a new and impressive look and feel. Thank you very much for this incredible effort! 🚀 

I have some minor non-blocking comments though:
- How can other plugins benefit from the new design without much refactoring? Can we improve the adoption by simply changing existing classes rather than adding new ones? 
- Would it make sense to describe the required steps for plugins (to behave correctly) somewhere in the developer documentation? This would help to get a consistent user interface for the whole area of Jenkins. (This can be done in a followup PR)
- How can we combine these changes with the existing Bootstrap and Datatables components so that everything is using the same styling? (Example: https://ci.jenkins.io/job/Plugins/job/warnings-ng-plugin/job/master/lastBuild/cpd)

File: core/src/main/resources/lib/layout/ionicon.jelly
Comment (timja) [reply core/src/main/resources/lib/layout/ionicon.jelly:17]: it's explained in the description, it's to avoid having to modify them into sprites.

File: war/src/main/less/abstracts/theme.less
Comment (janfaracik) [reply war/src/main/less/abstracts/theme.less:172]: I have! Might change a little but I've opened a PR - https://github.com/jenkinsci/dark-theme-plugin/pull/165 :)

**Review by timja — APPROVED** [review]
Going through here I can't see any blockers,  some questions that would be nice to have answers to, but we can iterate all of it.

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (janfaracik) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:71]: The SVGs should be hidden due to aria-hidden, so I've added the title label in case the SVG represented something important that wasn't supported by text.

File: core/src/main/resources/lib/layout/ionicon.jelly
Comment (janfaracik) [reply core/src/main/resources/lib/layout/ionicon.jelly:17]: Yep! The parent element would manage the styling for the icon, although modifier classes should continue to work.

File: core/src/main/resources/hudson/views/JobColumn/column.jelly
Comment (janfaracik) [reply core/src/main/resources/hudson/views/JobColumn/column.jelly:29]: There aren't existing classes for a link inside of a table afaik, and styling `.jenkins-table a` could have unintended consequences where a developer wants a link that doesn't look like `.jenkins-table__link`.

File: war/src/main/resources/images/ionicons/ellipse-outline.svg
Comment (janfaracik) [reply war/src/main/resources/images/ionicons/ellipse-outline.svg:1]: It'd involve copying the icons from ionicons to this folder, wanted to avoid that in this PR as it'd balloon this PR up in size.

**Comment by uhafner** [comment]
I'm not sure if you are looking into Jira as well: I created a bug report regarding the health-report tooltip styling:
https://issues.jenkins.io/browse/JENKINS-67154

**Comment by uhafner** [comment]
Can you please file an issue in Jira so we can track this problem?

File: war/src/main/resources/images/ionicons/ellipse-outline.svg
Comment (uhafner) [reply war/src/main/resources/images/ionicons/ellipse-outline.svg:1]: The icons are currently part of resources folder (and not below the webapp folder). How can we replace icons like [in actions](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/management/ConfigureLink.java) with these new ionicon icons?

Comment (timja) [reply war/src/main/resources/images/ionicons/ellipse-outline.svg:1]: We can't yet, it wasn't part of this feature.

likely the l:icon tag needs to be able to detect its an ionicon and load it, possibly another flag or other way

**Comment by didiez** [comment]
Is there any way to make the new UI more compact other than overriding with a custom css? IMHO The new "pills" aproach added too much vertical spacing for an operational UI.
Also, we have noticed in a quick check:

- the icon sizes S/M/L are not being respected anymore (16/24/32px)
- there is no "overlay panel" arrow for the build numbers on "hover"

**Comment by timja** [comment]
Can you create (at least one if not more) issue(s) please with more details, screenshots would be helpful, I understand some of what you mean but not all.

You'll need custom css or an improvement PR here.

**Comment by didiez** [comment]
> Can you create (at least one if not more) issue(s) please with more details, screenshots would be helpful, I understand some of what you mean but not all.

Sure :+1:
- https://issues.jenkins.io/browse/JENKINS-67395
- https://issues.jenkins.io/browse/JENKINS-67396

**Comment by daniel-beck** [comment]
Caused https://issues.jenkins.io/browse/JENKINS-67515

**Comment by janfaracik** [comment]
> Caused https://issues.jenkins.io/browse/JENKINS-67515

Fixed [here](https://github.com/jenkinsci/jenkins/pull/6162) :)

File: war/src/main/less/modules/table.less
Comment (KalleOlaviNiemitalo) [inline war/src/main/less/modules/table.less:119]: This `inline-flex` may have caused [JENKINS-68205](https://issues.jenkins.io/browse/JENKINS-68205).

File: war/src/main/less/modules/table.less
Comment (Bananeweizen) [inline war/src/main/less/modules/table.less:63]: This leads to _any_ table using a height of 50 px per row, while you probably wanted this only for the table listing the Jenkins jobs. For instance the test results table looks horrible with this huge space now.

Comment (NotMyFault) [reply war/src/main/less/modules/table.less:63]: Please file an appropriate issue labeled as regression on Jira (https://issues.jenkins.io/secure/Dashboard.jspa) including steps to reproduce and screenshots from the old and current behavior.

**Comment by basil** [comment]
Causes [JENKINS-68303](https://issues.jenkins.io/browse/JENKINS-68303).

**Comment by alecharp** [comment]
Causes [JENKINS-68402](https://issues.jenkins.io/browse/JENKINS-68402).

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (daniel-beck) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:71]: Caused [SECURITY-2761](https://www.jenkins.io/security/advisory/2022-06-22/#SECURITY-2781)

**Pull request merged**
